aws/service.ts

```diff
@@ -247,74 +247,109 @@ function computeImage(
-                pulumi.log.debug(`    already built: ${imageName} (${d})`, repository));
-        } else {
-            // If we haven't, build and push the local build context to the ECR repository, wait for
-            // that to complete, then return the image name pointing to the ECT repository along
-            // with an environment variable for the image digest to ensure the TaskDefinition get's
-            // replaced IFF the built image changes.
-            const {repositoryUrl, registryId} = repository!;
-            imageDigest = docker.buildAndPushImage(imageName, container.build, repositoryUrl, repository, async () => {
-                // Construct Docker registry auth data by getting the short-lived authorizationToken from ECR, and
-                // extracting the username/password pair after base64-decoding the token.
-                //
-                // See: http://docs.aws.amazon.com/cli/latest/reference/ecr/get-authorization-token.html
-                if (!registryId) {
-                    throw new RunError("Expected registry ID to be defined during push");
-                }
-                const credentials = await aws.ecr.getCredentials({ registryId: registryId });
-                const decodedCredentials = Buffer.from(credentials.authorizationToken, "base64").toString();
-                const [username, password] = decodedCredentials.split(":");
-                if (!password || !username) {
-                    throw new RunError("Invalid credentials");
-                }
-                return {
-                    registry: credentials.proxyEndpoint,
-                    username: username,
-                    password: password,
-                };
-            });
-            if (imageName) {
-                buildImageCache.set(imageName, imageDigest);
-            }
-            imageDigest.apply(d =>
-                pulumi.log.debug(`    build complete: ${imageName} (${d})`, repository));
-        }
-
-        preEnv.IMAGE_DIGEST = imageDigest;
-
-        return pulumi.all([repository.repositoryUrl, pulumi.all(preEnv)])
-                     .apply(([url, e]) => ({ image: url, environment: e }));
+        return computeImageFromBuild(preEnv, container.build, imageName, repository);
     }
     else if (container.image) {
-        return pulumi.all(preEnv).apply(e => ({ image: imageName, environment: e }));
+        return computeImageFromImage(preEnv, imageName);
     }
     else if (container.function) {
-        const func = container.function;
-        preEnv.PULUMI_SRC = pulumi.runtime.serializeFunctionAsync(func);
-
-        // TODO[pulumi/pulumi-cloud#85]: Put this in a real Pulumi-owned Docker image.
-        // TODO[pulumi/pulumi-cloud#86]: Pass the full local zipped folder through to the container (via S3?)
-        return pulumi.all(preEnv).apply(e => ({ image: imageName, environment: e }));
+        return computeImageFromFunction(container.function, preEnv, imageName);
     }
```

Comment:
i extracted everything into helpers to keep the overall logic easier to understand (IMO).  Container is effectively an Enum, so it helped to just have each enum case handled separately.